src/tree.rs

```diff
@@ -102,26 +293,34 @@ impl fmt::Display for Node {
     }
 }
 
-/// A node in a merged bookmark tree. Holds the local node, remote node,
-/// merged children, and a merge state indicating which side to prefer.
+/// Synced item kinds. Each corresponds to a Sync record type.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Kind {
+    Bookmark,
+    Query,
+    Folder,
+    Livemark,
+    Separator,
+}
+
+impl fmt::Display for Kind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(self, f)
+    }
+}
+
+/// A merged bookmark node that indicates which side to prefer, and holds merged
+/// child nodes.
+#[derive(Debug)]
 pub struct MergedNode<'t> {
     pub guid: String,
-    pub local_node: Option<&'t Node>,
-    pub remote_node: Option<&'t Node>,
-    pub merge_state: MergeState,
+    pub merge_state: MergeState<'t>,
     pub merged_children: Vec<Box<MergedNode<'t>>>,
```

Comment:
I don't think these need to be boxed (`mem::size_of::<Vec<T>>()` does not depend on `mem::size_of::<T>()`)